Dialtone V6 Component Avatar Implementation

ryandialpad opened the pull request:
# Dialtone V6 Component Avatar Implementation

<!--- Feel free to remove any unused sections -->

## :hammer_and_wrench: Type Of Change

<!--- Tick or place an `x` in all of the checkboxes that apply -->

- [ ] Fix
- [x] Feature
- [x] Refactoring
- [ ] Documentation

## :book: Description

<!--- Describe the changes -->

Created avatar css component
Created initial avatar documentation

## :bulb: Context

<!--- Describe the purpose of the changes -->
<!--- Why did we make these changes? -->
<!--- What problem(s) do they solve? -->

Closes #210 

## Pull Request Checklist

 - [x] Ask the contributors if a similar effort is already in process or has been solved.
 - [x] Review the [contribution guidelines](https://github.com/dialpad/dialtone/blob/staging/.github/CONTRIBUTING.md).
 - [x] Ensure all `gulp` scripts successfully compile.
 - [x] Update, remove, or extend all affected documentation.
 - [x] Provide a description what is being changed, extended, or removed.
 - [x] Link to any issue(s) this request is resolving.
 - [x] Ensure no private Dialpad links or info are in the code or pull request description (Dialtone is a public repo!).
 - [x] Request a review from Drew Chandler, Joshua Hynes, or Ted Goas.

## :camera: Screenshots / GIFs

<!--- Mandatory for any UI work -->
<!--- Link any screenshots / GIFs below -->

![Screen Shot 2021-03-29 at 11 16 37 AM](https://user-images.githubusercontent.com/73855198/112858980-69ca2080-9080-11eb-934a-034584d4c4f0.png)
![Screen Shot 2021-03-29 at 11 16 44 AM](https://user-images.githubusercontent.com/73855198/112858987-6a62b700-9080-11eb-9bb1-a6d43384cce7.png)
![Screen Shot 2021-03-29 at 11 16 57 AM](https://user-images.githubusercontent.com/73855198/112858997-6b93e400-9080-11eb-83aa-47d0ab652b40.png)
![Screen Shot 2021-03-29 at 11 17 05 AM](https://user-images.githubusercontent.com/73855198/112859002-6cc51100-9080-11eb-96a8-5a31a193084b.png)
![Screen Shot 2021-03-29 at 11 17 11 AM](https://user-images.githubusercontent.com/73855198/112859008-6df63e00-9080-11eb-8e18-40dcc13a727f.png)

dialpad-drew commented on lib/less/variables/spacing.less:
```diff
@@ -37,8 +39,10 @@
     su8:    @su8;
     su12:   @su12;
     su16:   @su16;
+    su18:   @su18;
```
As mentioned in another comment, there are a couple one-off values here that needed to be used to keep avatar visually consistent among its own variants. `18px` is one of those values. 

While I like the idea of keeping globals up-to-date so things don't get out of hand, at the same time, doesn't adding something to the globals open it up for use anywhere? What if that's not the intention? Should there be a threshold for adding things to our globals and, if so, should it be higher than a single occurrence?

That said, I'm having regrets about using 40px instead of 48px for the largest avatar. Can we use 48px instead? I'll make the update in Figma.

dialpad-drew commented on docs/components/avatars.html:
```diff
@@ -0,0 +1,138 @@
+      {% codeWellFooter %}
+        {% highlight html linenos %}
+<div class="d-avatar d-avatar--{{ size }}">
+  <img src="/{{ site.paths.img }}/person.png" alt="An avatar of a person" />
```
I think the image is disproportionately stretched wider than it is tall.
![Screen Shot 2021-03-29 at 7 05 20 PM](https://user-images.githubusercontent.com/77751297/112918272-c13cb080-90c1-11eb-8c02-d5bb9582a2c2.png)

ryandialpad replied:
I figured I would leave this one for later, I am not great at resizing / formatting images 🙃

dialpad-drew commented on docs/components/avatars.html:
```diff
@@ -0,0 +1,138 @@
+---
```
Nice job w/ the new documentation page! You dove right in like whoa! 🏊  🤿  🪂

ryandialpad replied:
Thanks!

dialpad-drew commented on lib/less/components/avatars.less:
```diff
@@ -0,0 +1,218 @@
+
+  img {
+    height:        var(--avatar--size);
+    width:         var(--avatar--size);
```
This might be your culprit on the image becoming disproportionate unless every image passed to this component will start with a 1:1 aspect ratio. If not, there will be some stretching.

Do we know how user-submitted images are currently handled? Are they constrained to 1:1 at the time of upload and/or before being passed to the UI? If so, this should probably be ok and we should just update the example image to be 1:1.

If the UI is passed non-1:1 images, we may need to consider another way.

ryandialpad replied:
I don't know off the top of my head how every image is passed. I have added styles to clip rather than stretch the images. Investigating now

ryandialpad replied:
Looks like our avatar images are 1:1 aspect ratio in the application. The following property `object-fit: cover` should catch any cases where the aspect ratio is not 1:1. It will simply attempt to cover the object without stretching the image

dialpad-drew commented on lib/less/components/avatars.less:
```diff
@@ -0,0 +1,218 @@
+  //  Component CSS Vars
+  //  ------------------------------------------------------------------------
+  --avatar--fc:           var(--muted-color);
+  --avatar--bgc:          transparent;
```
In the event an image doesn't load for reasons, would a transparent background color essentially lead to a hole in the UI where the avatar should have been?

ryandialpad replied:
Will investigate

ryandialpad replied:
Did you want to present any default content in the event that an image is not provided?

dialpad-drew replied:
A background-color would act as a fallback, right? The image should cover it but if the image fails to load, you at least have a gray circle (probably with a browser-provided broken image icon). If so, we could try `black-200`.

dialpad-drew commented:
In fact, now that I'm looking at it, `8px` font size for the `xs` is just way too friggin' small. I've made some changes I hope will reduce some of the one-offs and the "small-ness."

I hope this helps make things a little tighter. Let me know.

## Changes
### Icon sizes
* `lg` 48px (changed from 40)

### Initial font sizes
Note: All are now one initial instead of two
* `xs`: 11px (changed from 8px)
* `sm`: 16px (changed from 12px)
* `md`: 20px (changed from 16px)
* `lg`: 32px (changed from 20px)

### Removed "custom" icon color
I think this is just a Figma thing for designers; let's remove it

ryandialpad commented:
Removed added variables. Updated avatar sizes, icon sizes and font sizes based on feedback. Removed one off green color variant.

Investigating avatar image styles

ryandialpad commented:
@dialpad-drew Could you please review the sm and md initial variants as they look a little off (font size too large for the avatar or avatar too small for the font size)
